### PR TITLE
KAFKA-2585; ConsoleConsumer should not hang infinitely upon exception

### PR DESCRIPTION
**Pull request by lindong28**

### CODE REVIEW

**Comment by lindong28** [comment]
Due to imcompatibility beween Java 1.7 and Java 1.8, if Kafka ConsoleConsumer was compiled by Java 1.8 but run by Java 1.7, and if ConsoleConsumer has consumer.timeout.ms in the consumer.properties, then when it timesout, it will throw the following exception and hang infinitely.

This will cause problem for e.g. Ducktape system test, which currently runs consumer with Java 1.7 and waits for consumer to timeout. This bug causes the Ducktape to hang for a long time.

To prevent ConsoleConsumer to hang in case of such exception, we can wrap try/catch around run() and do System.exit(0) at the end of main() in ConsoleConsumer.

**Comment by lindong28** [comment]
Here is the exception.

[2015-09-28 05:41:31,499] ERROR Error processing message, stopping consumer:  (kafka.tools.ConsoleConsumer$)
kafka.consumer.ConsumerTimeoutException
    at kafka.consumer.ConsumerIterator.makeNext(ConsumerIterator.scala:69)
    at kafka.consumer.ConsumerIterator.makeNext(ConsumerIterator.scala:33)
    at kafka.utils.IteratorTemplate.maybeComputeNext(IteratorTemplate.scala:66)
    at kafka.utils.IteratorTemplate.hasNext(IteratorTemplate.scala:58)
    at kafka.utils.IteratorTemplate.next(IteratorTemplate.scala:38)
    at kafka.consumer.ConsumerIterator.next(ConsumerIterator.scala:46)
    at kafka.consumer.OldConsumer.receive(BaseConsumer.scala:70)
    at kafka.tools.ConsoleConsumer$.process(ConsoleConsumer.scala:94)
    at kafka.tools.ConsoleConsumer$.run(ConsoleConsumer.scala:57)
    at kafka.tools.ConsoleConsumer$.main(ConsoleConsumer.scala:41)
    at kafka.tools.ConsoleConsumer.main(ConsoleConsumer.scala)
Processed a total of 78 messages
Exception in thread "main" java.lang.NoSuchMethodError: java.util.concurrent.ConcurrentHashMap.keySet()Ljava/util/concurrent/ConcurrentHashMap$KeySetView;
    at kafka.utils.Pool.keys(Pool.scala:77)
    at kafka.consumer.FetchRequestAndResponseStatsRegistry$.removeConsumerFetchRequestAndResponseStats(FetchRequestAndResponseStats.scala:69)
    at kafka.metrics.KafkaMetricsGroup$.removeAllConsumerMetrics(KafkaMetricsGroup.scala:189)
    at kafka.consumer.ZookeeperConsumerConnector.shutdown(ZookeeperConsumerConnector.scala:202)
    at kafka.consumer.OldConsumer.stop(BaseConsumer.scala:75)
    at kafka.tools.ConsoleConsumer$.process(ConsoleConsumer.scala:98)
    at kafka.tools.ConsoleConsumer$.run(ConsoleConsumer.scala:57)
    at kafka.tools.ConsoleConsumer$.main(ConsoleConsumer.scala:41)
    at kafka.tools.ConsoleConsumer.main(ConsoleConsumer.scala)

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala [hunk 1]
```diff
@@ -38,7 +38,13 @@ object ConsoleConsumer extends Logging {
...
   def main(args: Array[String]) {
     val conf = new ConsumerConfig(args)
-    run(conf)
+    try {
+      run(conf)
+    } catch {
+      case e: Throwable =>
+        error("Error running consumer: ", e)
+    }
+    System.exit(0);
```
Comment (ijuma) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: Why is this needed? Are we leaking non-daemon threads? Also, we should probably exit with `System.exit(-1)` when there is an error.

Comment (lindong28) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: @ijuma Yes, I think we are leaking non-daemon threads here. I find that, if consumer is compiled with Java 7 and run with Java 8, when it times out, the above exception is wrong and consumer hang. To fix the problem, I need to both catch the exception and do System.exit(..).

I followed the practice used in ProducerPerformance.scala, which invokes `System.exit(0)` at the end of `main()`. Do you like me to add an additional System.exit(-1) right after `error(...)` in `catch{...}`, and remove `the System.exit(0)`?

Comment (ijuma) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: Yes, that would be better as we would return a non-zero code which some command-line tools may rely on. I would also suggest removing the braces for the `try` as they are not needed for a single expression `try`.

Comment (lindong28) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: I am not sure if I should remove the brace around `try`. Is this the recommended style for scala code in Kafka? I am asking because braces are added for single expression `try` in most parts of Kafka code, e.g. SocketServerTest.scala, Log.scala, ZkUtils.scala. I find it a bit confusing if `catch` has braces while try doesn't -- it only happened once in ConsoleConsumer.scala. What do you think?

Also, how about I use `System.exit(1)`? I find that most Kafka tools use `1` as error code. MirrorMaker is the only case which does `System.exit(-1)`.

Comment (ijuma) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: Having looked, I don't think the exception case is specified one way or another. What is specified is that braces should be skipped for single-expression if/else blocks. The reason why I think it should be removed is that it adds noise and it doesn't help in any way (if you add a new expression without adding the braces, there will be a compiler error due to the catch being in an unexpected place, so no safety problems). The reason why I think the existing code is the way you say is probably that many people are more familiar with Java than Scala. The one example you mention is actually quite recent and it was my suggestion. :) Anyway, it's up to you, since this is an area that is not well specified.

`System.exit(1)` is fine. I prefer `-1` as it's clearer to me that it's an error, but as you say it looks like Kafka code tends to use `1`.

Comment (lindong28) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: @ijuma Thanks much for your comment. I have spent time thinking through the pros and cons of having braces. I finally decided to keep the braces, because 
- it doesn't add any additional line as compared to not having the brace
- the code style will be consistent with other parts of the Kafka code. It can be a little confusing if you have different code style even in the same file, e.g. ConsoleConsumer.

Please see updated patch. I have tested again to make sure the code solves the problem. Thanks!

File: config/consumer.properties [hunk 2]
```diff
@@ -17,13 +17,13 @@
...
-zookeeper.connection.timeout.ms=6000
+#zookeeper.connection.timeout.ms=6000
 
 #consumer group id
-group.id=test-consumer-group
+#group.id=test-consumer-group
 
 #consumer timeout
-#consumer.timeout.ms=5000
+consumer.timeout.ms=3000
```
Comment (ijuma) [inline config/consumer.properties]: I think these changes are unintentional.

Comment (lindong28) [inline config/consumer.properties]: Ah, my bad. I made these change only to test the patch. The change was reverted just now.

Thanks @ijuma.

**Comment by ijuma** [comment]
LGTM

**Comment by onurkaraman** [comment]
LGTM

**Comment by lindong28** [comment]
Hey @gwenshap , this patch has been reviewed by @ijuma and @onurkaraman. The patch addresses the Ducktae system test problem that is also reported in KAFKA-2573. Would you have time to review the patch? 

Thanks,

**Comment by gwenshap** [comment]
LGTM

**Comment by lindong28** [comment]
Thank you @gwenshap.
